File: 백준/Gold/11000. 강의실 배정/강의실 배정.py
import sys
from heapq import heappush, heappop


def solution():
    n = int(sys.stdin.readline())
    arr = []

    for _ in range(n):
        start, end = map(int, sys.stdin.readline().split())
        arr.append([start, end])
    arr.sort()

    room = []
    heappush(room, arr[0][1])

    for i in range(1, n):
        if arr[i][0] >= room[0]:
            heappop(room)
            heappush(room, arr[i][1])
        else:
            heappush(room, arr[i][1])
    print(len(room))

solution()

File: 백준/Gold/11000. 강의실 배정/강의실 배정.java
import java.util.*;
import java.lang.*;
import java.io.*;

// The main method must be in a class named "Main".
class Main {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());

        int[][] arr = new int[n][2];

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            int start = Integer.parseInt(st.nextToken());
            int end = Integer.parseInt(st.nextToken());

            arr[i][0] = start;
            arr[i][1] = end;
        }

        Arrays.sort(arr, (a, b) -> {
            if (a[0] == b[0]) return a[1] - b[1];
            return a[0] - b[0];
        });

        // System.out.println(Arrays.deepToString(arr));
        
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> {
            if (a[1] == b[1]) return Integer.compare(a[0], b[0]);
            return Integer.compare(a[1], b[1]);
        });

        for (int[] a : arr) {
            if (pq.isEmpty()) {
                pq.add(a);
                continue;
            }

            int[] min_arr = pq.peek();
            if (min_arr[1] <= a[0]) pq.poll();

            pq.add(a);
        }

        System.out.println(pq.size());

    }
}